lib/redirect-pool.js

```diff
@@ -0,0 +1,187 @@
+  }
+
+  for (let i = 0; i < headers.length; i += 2) {
+    // Find the matching headers, then return its value
+    if (headers[i].length && headers[i].toLowerCase() === 'location') {
+      return headers[i + 1]
+    }
+  }
+
+  return null
+}
+
+// https://tools.ietf.org/html/rfc7231#section-6.4.4
+function shouldRemoveHeader (header, removeContent) {
+  const lcHeader = header.toLowerCase()
+
+  return lcHeader === 'host' || (removeContent && lcHeader.indexOf('content-') === 0)
+}
+
+// https://tools.ietf.org/html/rfc7231#section-6.4
+function cleanRequestHeaders (headers, removeContent) {
+  if (Array.isArray(headers)) {
+    for (let i = headers.length - 2; i >= 0; i -= 2) {
+      if (shouldRemoveHeader(headers[i], removeContent)) {
+        headers.splice(i, 2)
+      }
+    }
+  } else {
+    // IncomingHttpHeaders
+    for (const header of Object.keys(headers)) {
+      if (shouldRemoveHeader(header, removeContent)) {
+        headers[header] = undefined
+      }
+    }
+  }
+}
+
+class RedirectPool extends Pool {
+  constructor (origin, options) {
+    super(origin, options)
+
+    this[kOrigin] = origin
+  }
+
+  dispatch (opts, handler) {
+    const pool = this
+
+    // Cannot use RedirectPool on pipelines
+    if (handler.handler) {
+      throw new InvalidArgumentError('RedirectPool cannot be used with pipeline')
+    }
+
+    if (util.isStream(opts.body)) {
+      throw new InvalidArgumentError('body cannot be a stream when using RedirectPool')
+    }
+
+    const maxRedirections = parseMaxRedirections(opts)
+
+    const redirectingHandler = {
+      onConnect (abort) {
+        handler.onConnect(abort)
+      },
+
+      onUpgrade (statusCode, headers, socket) {
+        handler.onUpgrade(statusCode, headers, socket)
+      },
+
+      onError (error) {
+        handler.onError(error)
+      },
+
+      onHeaders (statusCode, headers, resume) {
+        let { [kLeftRedirections]: leftRedirections, [kFollowedRedirections]: followedRedirections } = opts
+
+        if (typeof leftRedirections !== 'number') {
+          leftRedirections = maxRedirections
+        }
+
+        // Check if statusCode is 3xx, if there is a location header and if the redirection can be followed
+        const location = redirectLocation(statusCode, headers, leftRedirections)
+
+        // Nothing to follow, use the original implementation
+        if (!location) {
+          return handler.onHeaders(statusCode, headers, resume, { redirections: followedRedirections })
+        }
+
+        // Gather other request options that will be modified
+        let { method, headers: requestHeaders, body } = opts
+
+        /*
+          https://tools.ietf.org/html/rfc7231#section-6.4
+
+          Remove headers referring to the original URL.
+          By default it is Host only, unless it's a 303 (see below), which removes also all Content-* headers.
+        */
+        if (requestHeaders) {
+          cleanRequestHeaders(requestHeaders, statusCode === 303)
```

Comment:
This mutates user owned object

Reply:
Good catch. Fixed.

Reply:
Did you push? I don't see a commit?

Reply:
Apparently I forgot. :)

I pushed right now!